### ISSUE TITLE
A3C Crashes on training with Comp Graph

### ISSUE THREAD

**Issue by Ailanz** [deeplearning4j/rl4j#86]
#### Issue Description
Hi, this is related to the recent fix that A3C networks requires a probability output (Softmax).

when running the A3C with

```
    private static ActorCriticFactoryCompGraphStdDense.Configuration NET_A3C =  ActorCriticFactoryCompGraphStdDense.Configuration
            .builder().l2(0.01).numHiddenNodes(20).useLSTM(false).numLayer(2).build();

```
we get:

```
java.lang.RuntimeException: Output from network is not a probability distribution
	at org.deeplearning4j.rl4j.policy.ACPolicy.nextAction(ACPolicy.java:65)
	at org.deeplearning4j.rl4j.policy.ACPolicy.nextAction(ACPolicy.java:20)
	at org.deeplearning4j.rl4j.learning.async.AsyncThreadDiscrete.trainSubEpoch(AsyncThreadDiscrete.java:77)
	at org.deeplearning4j.rl4j.learning.async.AsyncThread.run(AsyncThread.java:82)
```

It doesn't happen every run but about 30% of the run, it runs into this error after training for a bit.

This is from the latest Master branch.
full code to reproduce:

```
package rl;

import org.deeplearning4j.rl4j.learning.async.a3c.discrete.A3CDiscrete;
import org.deeplearning4j.rl4j.learning.async.a3c.discrete.A3CDiscreteDense;
import org.deeplearning4j.rl4j.mdp.toy.SimpleToy;
import org.deeplearning4j.rl4j.mdp.toy.SimpleToyState;
import org.deeplearning4j.rl4j.network.ac.ActorCriticFactoryCompGraphStdDense;
import org.deeplearning4j.rl4j.util.DataManager;

import java.io.IOException;

public class Toy {

    public static void main(String[] args) throws IOException {
        simpleToy();
    }

    public static void simpleToy() throws IOException {
        DataManager manager = new DataManager();
        SimpleToy mdp = new SimpleToy(20);
        int in = mdp.getObservationSpace().getShape()[0];
        A3CDiscreteDense<SimpleToyState> dql = new A3CDiscreteDense<>(mdp, NET_A3C, A3C_CONFIG, manager);
        dql.train();
        mdp.close();
    }

    private static A3CDiscrete.A3CConfiguration A3C_CONFIG =
            new A3CDiscrete.A3CConfiguration(
                    123,            //Random seed
                    800000,            //Max step By epoch
                    50000,         //Max step
                    12,              //Number of threads
                    5000000,              //t_max
                    0,             //num step noop warmup
                    1,           //reward scaling
                    0.99,           //gamma
                    10           //td-error clipping
            );

    private static ActorCriticFactoryCompGraphStdDense.Configuration NET_A3C =  ActorCriticFactoryCompGraphStdDense.Configuration
            .builder().l2(0.01).numHiddenNodes(20).useLSTM(false).numLayer(2).build();
    
}

```

Please describe your issue, along with:
- expected behavior
- encountered behavior

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version
* platform information (OS, etc)
* CUDA version, if used
* NVIDIA driver version, if in use

#### Contributing

If you'd like to help us fix the issue by contributing some code, but would
like guidance or help in doing so, please mention it!

**Comment by saudet** [comment]
How did you run this example? I'm getting a different problem with 0.9.2-SNAPSHOT:
```java
org.nd4j.linalg.exception.ND4JIllegalStateException: Op name subtract is missing!
	at org.nd4j.linalg.api.ops.DynamicCustomOp.opHash(DynamicCustomOp.java:258)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1533)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.subi(BaseNDArray.java:3255)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.subi(BaseNDArray.java:3225)
	at org.nd4j.linalg.lossfunctions.impl.LossL2.computeGradient(LossL2.java:119)
	at org.nd4j.linalg.lossfunctions.impl.LossMSE.computeGradient(LossMSE.java:56)
	at org.deeplearning4j.nn.layers.BaseOutputLayer.getGradientsAndDelta(BaseOutputLayer.java:173)
	at org.deeplearning4j.nn.layers.BaseOutputLayer.backpropGradient(BaseOutputLayer.java:148)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.calcBackpropGradients(MultiLayerNetwork.java:1357)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.backprop(MultiLayerNetwork.java:1307)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2311)
	at org.deeplearning4j.rl4j.network.ac.ActorCriticSeparate.gradient(ActorCriticSeparate.java:81)
	at org.deeplearning4j.rl4j.learning.async.a3c.discrete.A3CThreadDiscrete.calcGradient(A3CThreadDiscrete.java:106)
	at org.deeplearning4j.rl4j.learning.async.a3c.discrete.A3CThreadDiscrete.calcGradient(A3CThreadDiscrete.java:28)
	at org.deeplearning4j.rl4j.learning.async.AsyncThreadDiscrete.trainSubEpoch(AsyncThreadDiscrete.java:123)
	at org.deeplearning4j.rl4j.learning.async.AsyncThread.run(AsyncThread.java:82)
```
@raver119 ? @AlexDBlack ?

**Comment by Ailanz** [comment]
Hi, I ran into both of these issues as well. But they do tend to go away after running it for a couple of runs and / or rebuild rl4j from maven.

**Comment by saudet** [comment]
Ok, when it does run I get the following output:
```
java.lang.RuntimeException: Output from network is not a probability distribution: [   �,     �]
	at org.deeplearning4j.rl4j.policy.ACPolicy.nextAction(ACPolicy.java:65)
	at org.deeplearning4j.rl4j.policy.ACPolicy.nextAction(ACPolicy.java:20)
	at org.deeplearning4j.rl4j.learning.async.AsyncThreadDiscrete.trainSubEpoch(AsyncThreadDiscrete.java:77)
	at org.deeplearning4j.rl4j.learning.async.AsyncThread.run(AsyncThread.java:82)
```
Based on the latest commit: https://github.com/deeplearning4j/rl4j/commit/4fbeac8b3b9354133c25d76adf6c65712e4ad505

So we're just getting NaNs here. You'll need to adjust the network parameters so this doesn't happen: https://deeplearning4j.org/troubleshootingneuralnets

**Comment by Ailanz** [comment]
ok thank you. Will do. If you guys could look into the other 2 earlier errors that would be swell as well!

Thanks!

**Comment by Ailanz** [comment]
Thanks! Will do!

**Comment by Ailanz** [comment]
Given that alot of the paremeters were pre-configured:
```
    private static ActorCriticFactoryCompGraphStdDense.Configuration NET_A3C =  ActorCriticFactoryCompGraphStdDense.Configuration
            .builder().l2(0.01).numHiddenNodes(20).useLSTM(false).numLayer(2).build();
```

and


```
    private static A3CDiscrete.A3CConfiguration A3C_CONFIG =
            new A3CDiscrete.A3CConfiguration(
                    123,            //Random seed
                    800000,            //Max step By epoch
                    50000,         //Max step
                    12,              //Number of threads
                    5000000,              //t_max
                    0,             //num step noop warmup
                    1,           //reward scaling
                    0.99,           //gamma
                    10           //td-error clipping
            );
```

do you have any tips to prevent the Nans? On that page you provided, I tried to increase reward factor to 10, 100, and 1000 (hopefully to prevent underflow) but it still throws the same error of Nans,

Any advice?

Thanks!

**Comment by saudet** [comment]
> The learning rate is one of, if not the most important hyperparameter.

Please start reading...

**Comment by Ailanz** [comment]
Silly question, since I've updated to 0.9.2-SNAPSHOT, I noticed that the setter for learning rate is no longer accessible via:
```
  private static ActorCriticFactoryCompGraphStdDense.Configuration NET_A3C =  ActorCriticFactoryCompGraphStdDense.Configuration
            .builder().numHiddenNodes(120).useLSTM(false).numLayer(2).build();
```

what is an alternative way to set the learning rate on my network.

Thanks!

**Comment by saudet** [comment]
It is part of the Updater now. Check the updated examples:
https://github.com/deeplearning4j/dl4j-examples/tree/sa_snapshot/rl4j-examples/src/main/java/org/deeplearning4j/examples/rl4j
